### ISSUE TITLE
Data loaders from remote URLs 

### ISSUE THREAD

**Issue by talmo** [talmolab/sleap-io#51]
It would be super convenient to support loading from remote data sources like URLs, Google Drive links, and cloud buckets (especially in notebooks). This would basically just automate downloading into a temp or local directory before calling the usual loading routines.

**Comment by talmo** [comment]
## Design proposal: remote URL loaders for `.slp`, `pkg.slp`, and (eventually) raw video

I spent some time investigating this end-to-end. The original framing was "download to a temp dir, then call the usual loading routines," but the empirical work showed we can do quite a bit better — most notably, load a 77 MB `pkg.slp` file from a URL while only pulling **~7.4 MB / 9.4% of the file in 8 HTTP requests**, faster than the full-download approach. That changes the UX meaningfully (notebooks become responsive instead of waiting on cold downloads), so I want to lay out the design before opening a PR.

This is long because there are several decisions worth documenting; happy to discuss any of them before implementation starts.

---

### TL;DR — what users will get

```python
import sleap_io as sio

# HTTPS — works out of the box (fsspec + aiohttp are base deps)
labels = sio.load_slp("https://slp.sh/4M16VD/labels.slp")

# Cloud schemes — single install enables all three
# pip install 'sleap-io[cloud]'   # brings s3fs + gcsfs + adlfs
labels = sio.load_slp("s3://my-bucket/path/labels.slp")
labels = sio.load_slp("gs://my-project/labels.slp")
labels = sio.load_slp("abfs://container@account.dfs.core.windows.net/labels.slp")

# Authenticated URLs
labels = sio.load_slp(
    "https://my-org.io/private/labels.slp",
    headers={"Authorization": "Bearer secret-token"},
)

# Iteration on the same URL across sessions → persistent cache
labels = sio.load_slp(
    "https://slp.sh/4M16VD/labels.slp",
    stream_mode="filecache",   # full DL, kept on disk, ETag-revalidated
    cache_expiry=86400,         # check once per day
)

# Polymorphic dispatch + magic-byte sniffing for ambiguous extensions
labels = sio.load_file("https://example.com/coco_annotations.json")
labels = sio.load_file("https://example.com/dlc_keypoints.h5")

# Cleanup
sio.clear_remote_cache(older_than=7 * 86400)
```

`Video.exists()` and embedded-video frame reads (the `pkg.slp` use case) also work transparently — the underlying HDF5 backend opens its own range-based reader, so a `pkg.slp` URL streams frames on demand without downloading the whole 77 MB file first.

---

### Backend choice

The load-bearing question is **which HTTP-range strategy** to use. Four candidates were considered:

| Approach | Bytes pulled (77 MB pkg) | HTTP requests | Wall (median of 5) | Verdict |
|---|---:|---:|---:|---|
| **`fsspec` HTTPFile with `cache_type='blockcache'`, `block_size=1 MiB`** | **7.4 MB (9.4%)** | **8** | **1.47 s** | **✅ Chosen** |
| `fsspec` HTTPFile with `cache_type='mmap'`, `block_size=1 MiB` | 7.4 MB (9.4%) | 8 | 1.41 s | Equivalent; mmap is harder to reason about on Windows |
| Hand-rolled `RangeFile` (httpx + LRU page cache, ~150 LOC) | 7.4 MB (9.4%) | 8 | 1.96 s | Same algorithm, slower (httpx vs aiohttp pooling), and we'd maintain it |
| `obstore` (Rust-backed, universal HTTP/S3/GCS/Azure) via `open_reader` | 0.44 MB (0.6%) | 32 | 6.02 s | `open_reader` has no page caching → many small requests; would need a custom `get_ranges` adapter (~50 LOC) to be competitive |
| Full HTTPS download (control) | 79 MB (100%) | 1 | 2.59 s | Baseline; range strategies beat it on wall + bandwidth |
| `fsspec` with default `cache_type` (sequential read-ahead) | 20.8 MB (26.2%) | 13 | 2.54 s | The wrong cache shape for h5py random access — included as a cautionary tale |

**Why `cache_type='blockcache'`:** h5py's HDF5 access pattern is small reads at scattered offsets (B-tree walks, superblock + footer, then chunk reads). `BytesCache` (fsspec's default, sequential read-ahead) discards already-fetched blocks on backward seeks → fetches the same metadata multiple times. `BlockCache` is an LRU over fixed-size blocks — exactly the right shape — and fsspec ships it natively.

Methodology: each backend × URL combo ran 1 warm-up trial (discarded) + 5 timed trials, with byte-counter assertions to catch silent-instrumentation failures. Tested against 3 URLs (43 KB, 1.2 MB, 77 MB) plus a multi-CDN probe across 5 hosts (slp.sh, AWS S3, GCS, GitHub release asset, HuggingFace blob); fsspec's blockcache strategy worked on every CDN without per-host tuning.

**Why not vendor:** identical bytes, identical requests, slower wall, ~150 LOC of HTTP client code we'd own forever. fsspec gets bug fixes, multi-cloud auth, and `simplecache::`/`filecache::` for free.

**Why not `obstore`:** genuinely interesting (Rust-backed, multi-range `get_ranges`, unified HTTP/S3/GCS/Azure surface). But the file-like API (`open_reader`) has no caching layer; using it for h5py means many tiny requests. To compete we'd write essentially the same LRU page cache as the vendored version, just using obstore's HTTP client. Not worth it today. Worth revisiting in 6-12 months when obstore's file-like API matures.

---

### Public API

```python
def load_slp(
    filename_or_url: str | os.PathLike,
    open_videos: bool = True,
    *,
    headers: dict[str, str] | None = None,
    stream_mode: Literal["auto", "blockcache", "cache", "filecache", "download"] = "auto",
    cache_storage: str | os.PathLike | None = None,
    cache_expiry: float | None = None,
    block_size: int = 1 << 20,
    max_blocks: int = 32,
    retries: int = 3,
) -> Labels:
    ...
```

For local paths, URL-specific kwargs are no-ops. `load_file(url, sniff=...)` and `load_video(url)` follow the same shape.

Top-level adds: `sleap_io.clear_remote_cache`, `sleap_io.RemoteIOError`.

**`stream_mode` semantics:**

| Mode | What it does | When to use | RAM | Disk | Repeat-open |
|---|---|---|---|---|---|
| `"auto"` (= `"blockcache"`) | Lazy HTTP Range reads with 32 × 1 MiB LRU page cache per file | **Default**; exploratory work, one-shot reads | ≤ 32 MiB/file | 0 | Cold |
| `"cache"` | `simplecache::` — full DL once, persistent | Iterating same URL across sessions | ~0 | `file_size` | Instant (no HTTP) |
| `"filecache"` | `filecache::` — full DL + TTL ETag revalidation | Want persistent + want to catch upstream updates | ~0 | `file_size` | HEAD per `cache_expiry` |
| `"download"` | Ephemeral full DL into `io.BytesIO` | One-off scripts, no disk cost | `file_size` (RAM) | 0 | Cold |

The threshold sweep showed no size-based crossover where another mode beats blockcache enough to justify auto-switching, so `auto` simply means blockcache. Users opt into the others.

**`RemoteIOError(OSError)`** wraps every `aiohttp.ClientError` + `fsspec` exception into a uniform type with `url` (always redacted — userinfo and token query params stripped), `status`, and `cause`. Mapping:

| Underlying | Maps to |
|---|---|
| `ClientResponseError(404)` | `RemoteIOError("file not found", status=404)` |
| `ClientResponseError(416)` | `RemoteIOError("range past EOF", status=416)` |
| `ClientResponseError(5xx)` | `RemoteIOError(generic, status=…)` (after retries) |
| `ClientConnectorError` | `RemoteIOError("connection error")` |
| `ClientPayloadError` | `RemoteIOError("truncated body")` |
| `TimeoutError` | `RemoteIOError("timeout")` |

Retries cover `{429, 500, 502, 503, 504}` with 200/400/800 ms exponential backoff and honor `Retry-After` (capped at 30 s).

---

### What's in the implementation

Single new internal module `sleap_io/io/_remote.py` (~250 LOC) holds:
- URL detection (`_is_url`) — robust against Windows paths (`C:\foo` → scheme `c`, not in allowlist)
- URL redaction (`_redact_url`) — strips userinfo, redacts token query params (`token`, `access_token`, `sig`, `sas`, etc.)
- `_build_fsspec_filesystem(scheme, *, headers=…)` — the single source of truth for fsspec configuration
- `_safe_get_client(...)` — `aiohttp.ClientSession` with a `TraceConfig.on_request_redirect` hook that strips `Authorization`/`Cookie`/`Proxy-Authorization` on cross-origin redirect (belt-and-suspenders; aiohttp 3.13.5+ does this natively but the hook protects against accidental downgrades)
- `open_url(url, *, stream_mode, …)` — dispatches by mode (blockcache / simplecache / filecache / download), returns a file-like
- `_sniff_format(url)` — 1-RTT `Range: bytes=0-15` magic-byte probe (HDF5 magic, `{`/`[` for JSON, ASCII+`,` for CSV)
- `clear_remote_cache(...)` — refuses to operate on `/`, `$HOME`, or directories without sleap-io's marker file; only deletes files matching fsspec's cache-key pattern
- `RemoteIOError`, `_raise_remote(e, url=…)` error mapping

Edits to existing modules:
- `sleap_io/io/main.py` — `load_slp(url)`, `load_file(url)`, `load_video(url)` (raises `NotImplementedError` pointing at a follow-up PR)
- `sleap_io/io/slp.py` — thread an optional already-open `h5py.File` through ~17 read helpers + the `read_labels` orchestrator (see "Helper refactor" below)
- `sleap_io/io/utils.py` — same pattern for the 3 generic `read_hdf5_*` helpers
- `sleap_io/io/video_reading.py` — `HDF5Video` URL-aware open + `__getstate__/__setstate__` (also fixes a latent pickle bug — see below)
- `sleap_io/model/video.py` — `Video.exists()` URL fast path with TTL cache
- `sleap_io/__init__.py` — re-export `clear_remote_cache`, `RemoteIOError`
- `pyproject.toml` — base deps + `[cloud]` extra

---

### The helper refactor — why and how

Without this, every `load_slp(url)` call opens the remote HDF5 file ~15 times (each read helper opens its own `h5py.File`), making fsspec parse HDF5 metadata afresh 15× per load. Wasteful even locally; painful on URLs.

**Pattern:** thread an optional already-open `h5py.File` via a keyword-only `_hdf5_file: h5py.File | None = None` kwarg. The convention already exists at one site — `make_video` at `sleap_io/io/slp.py:188` — so it gets adopted everywhere for symmetry.

```python
def read_tracks(labels_path, *, _hdf5_file=None):
    return read_hdf5_dataset(labels_path, "tracks_json", _hdf5_file=_hdf5_file)


def _read_labels_from_open_file(labels_path, f: h5py.File, *, open_videos):
    tracks = read_tracks(labels_path, _hdf5_file=f)
    videos = read_videos(labels_path, _hdf5_file=f, open_backend=open_videos)
    skeletons = read_skeletons(labels_path, _hdf5_file=f)
    # … 14 more …
```

The kwarg is keyword-only, underscore-prefixed (private), and defaults to `None`. Every existing positional call `read_hdf5_dataset(path, dataset)` continues to work without modification — verified by reading the test suite.

**Scope:** ~17 helpers in `slp.py` + 3 in `utils.py` + the orchestrator. ~220 LOC.

**One wrinkle:** `read_label_images` keeps a long-lived `h5py.File` open for lazy frame access (stored on `Labels._label_image_file`, already excluded from pickle). It uses `_hdf5_file=` for metadata reads but opens a separate handle for the lazy data — saves one open of two.

---

### URL-aware `HDF5Video` + a pickle bug fix

For `pkg.slp` files (embedded videos as HDF5 datasets), the `HDF5Video` backend opens its own fsspec file so embedded frame reads also go through `blockcache`. New `attrs` slots (slotted attrs classes require explicit field declarations — can't just `self.x = ...`):

```python
@attrs.define
class HDF5Video(VideoBackend):
    # existing fields …
    _url_file: object | None = attrs.field(
        init=False, default=None, repr=False, eq=False
    )
    _url_headers: dict[str, str] | None = attrs.field(
        init=False, default=None, repr=False, eq=False
    )
    _url_stream_mode: str = attrs.field(
        init=False, default="blockcache", repr=False, eq=False
    )

    def _open_h5(self) -> h5py.File:
        if _is_url(self.filename):
            if self._url_file is None:
                self._url_file = open_url(
                    self.filename, headers=self._url_headers,
                    stream_mode=self._url_stream_mode,
                )
            return h5py.File(self._url_file, "r")
        return h5py.File(self.filename, "r")
```

**Pickle bug, fixed as a side effect:** `HDF5Video` with `_open_reader != None` is currently un-picklable (h5py.File holds a thread lock that doesn't pickle). Latent bug — surfaces in PyTorch `DataLoader` workers and `multiprocessing.Pool`. The URL work brings it to the front, so we fix it via `__getstate__/__setstate__` on `VideoBackend` (drops `_open_reader`) and `HDF5Video` (also drops `_url_file`). Same pattern `Labels.__getstate__/__setstate__` already uses for `_label_image_file`.

```python
@attrs.define
class VideoBackend:
    def __getstate__(self) -> dict:
        import attr
        state = {a.name: getattr(self, a.name) for a in attr.fields(type(self))}
        state["_open_reader"] = None
        return state

    def __setstate__(self, state: dict) -> None:
        for key, value in state.items():
            object.__setattr__(self, key, value)
```

---

### `Video.exists()` — TTL cache because GUI hot path

GUIs (e.g. the SLEAP labeler) call `Video.is_open` on every render. For 30 videos at 60 FPS that's 1800 calls/sec; one HEAD request per call would melt the network.

The URL branch must be the **first statement** of `Video.exists()` — `is_file_accessible(url)` (the existing local check) does `open(url, "rb")` which raises and gets swallowed, short-circuiting to `False` for any URL. Patch: URL check first; if it's a URL, hit a per-instance TTL cache (60 s default, `$SLEAP_IO_EXISTS_TTL` to override) on a HEAD probe. HEAD with `Range: bytes=0-0` fallback handles servers that return 405 to HEAD (Cloudflare Workers, some Drive endpoints).

```python
_exists_cache: dict[tuple[str, str | None], tuple[bool, float]] = attrs.field(
    init=False, factory=dict, repr=False, eq=False
)
```

Cleared in `Video.replace_filename(...)`.

---

### Security requirements (binding)

A real concern with remote loading + authenticated URLs is leaking credentials on redirects. The probe across 5 CDNs found that **2 of 5 hosts (HuggingFace, GitHub) use cross-origin redirects to short-lived signed URLs** — i.e., this isn't hypothetical, it's the common case.

| # | Requirement | Verified by |
|---|---|---|
| **S1** | `aiohttp>=3.13.5` base dep | Full `Authorization`+`Cookie`+`Proxy-Authorization` stripping on cross-origin redirect (older aiohttp only strips `Authorization`; v3.13.5 added the other two — important if users pass session cookies) |
| **S2** | Belt-and-suspenders `TraceConfig.on_request_redirect` hook in `_safe_get_client` | `tests/io/test_remote.py::test_strips_auth_on_cross_origin_redirect` — set `Authorization: Bearer ...`, hit a 302 to a different origin, assert the redirect target's request lacks `Authorization` |
| **S3** | TLS `verify=True`; **no public `verify=False` kwarg** | Code review |
| **S4** | `_redact_url` in all logs, error messages, cache sidecars | `RemoteIOError.url` is always redacted |
| **S5** | `Accept-Encoding: identity` on every request | Avoids the gzip + Range offset-ambiguity trap |
| **S6** | `clear_remote_cache` refuses dirs lacking `.sleap-io-cache-marker` | Refuses `/`, `$HOME`, `~` even if marker present |
| **S7** | Cache files have restrictive POSIX permissions (`0600`) where possible | umask set before writes; POSIX only |
| **S8** | All aiohttp / fsspec exceptions wrapped as `RemoteIOError` | No raw stack traces with `Authorization` headers |

Headers can also surface in URL query strings (`?token=…`). `_redact_url` redacts `token`, `access_token`, `x-amz-security-token`, `sas`, `sig` query params.

---

### Dependencies

Base (always installed):

| Package | Floor | Why |
|---|---|---|
| `fsspec` | `>=2024.10.0` | `cache_type='blockcache'` + `cache_options={"max_blocks": …}` API stability |
| `aiohttp` | `>=3.13.5` | Full cross-origin sensitive-header stripping |

Transitively pulls `attrs`, `multidict`, `yarl`, `frozenlist`, `aiosignal`, `propcache`, `aiohappyeyeballs` — ~6 packages, ~3 MB. Most already present transitively from sleap-io's existing deps.

`[cloud]` extra — **one extra brings all three**:

```toml
cloud = ["s3fs>=2024.10.0", "gcsfs>=2024.10.0", "adlfs>=2024.10.0"]
```

`pip install 'sleap-io[cloud]'` lights up `s3://`, `gs://`/`gcs://`, `az://`/`abfs://` simultaneously. Using a cloud scheme without the extra raises `ImportError` with the install hint.

Dev: `pytest-httpserver>=1.0` for the security/edge-case test wave.

`[pyav]` extra is unchanged; remote video lands in a follow-up.

---

### Performance acceptance criteria (in CI)

Median of 5 trials of `load_slp("https://slp.sh/4M16VD/labels.slp")`:

| Metric | Threshold | Measured |
|---|---|---|
| Bytes pulled | ≤ 10 MB (≤ 13% of file) | 7.4 MB (9.4%) |
| HTTP requests | ≤ 10 | 8 |
| Wall time (US fast link) | ≤ 2.5 s | 1.47 s |
| RAM peak / open Video | ≤ 64 MiB | 32 MiB |

Any regression fails CI. Live URL test marked `@pytest.mark.live` so it can be skipped on offline runs; the bulk of tests use `pytest-httpserver`.

---

### Edge cases verified under controlled servers

Tested with `pytest-httpserver`:

| Edge case | fsspec result | Our wrapper |
|---|---|---|
| HEAD returns 405 | Falls back to ranged GET automatically | ✅ no extra code |
| Server returns 200 instead of 206 on Range | Silently accepts (works for small files; risky for large) | ⚠ Validate `Content-Range` matches request |
| Server gzips a binary body | aiohttp decompresses transparently (offsets become meaningless) | ⚠ Send `Accept-Encoding: identity` |
| Cross-origin redirect with `Authorization` set | Strips on aiohttp 3.13.5+ | ✅ Plus belt-and-suspenders hook |
| 416 Range Not Satisfiable | Raises cleanly | ✅ Map to `RemoteIOError` |
| Multipart byteranges | Rejected by 4/5 CDNs (slp.sh, GCS, GitHub/Azure, AWS); only HF/CloudFront supports | We don't emit them anyway |

---

### Implementation outline (subject to change during PR)

This is what the PR(s) will look like; the exact split + ordering may change once implementation starts.

Single-PR variant (~700 LOC impl + ~620 LOC tests + ~180 lines docs ≈ 1100 LOC total):

1. Add `pytest-httpserver` dev dep + scaffold `tests/io/test_remote.py`
2. Update `pyproject.toml` — base deps + `[cloud]` extra
3. Create `sleap_io/io/_remote.py`
4. Refactor `sleap_io/io/utils.py` — `_hdf5_file=` keyword-only kwarg
5. Refactor `sleap_io/io/slp.py` — thread `_hdf5_file=` through ~17 helpers + orchestrator
6. `HDF5Video` URL awareness + `VideoBackend`/`HDF5Video` pickle hooks
7. `load_slp(url)` in `main.py` — wire `open_url`, propagate config to embedded videos
8. `load_file(url)` — sniff dispatch
9. Security test wave (the S1–S8 tests above)
10. `Video.exists()` URL fast path + TTL cache
11. `load_video(url)` `NotImplementedError` + docs + top-level re-exports

Could also reasonably split into 2-3 PRs if the single large one is undesirable for review. The natural split points are:
- (a) `_remote.py` standalone + tests + `_hdf5_file=` helper refactor (no public API change yet)
- (b) `load_slp(url)` + `load_file(url)` + `HDF5Video` URL awareness
- (c) `Video.exists()` URL path + `load_video(url)` `NotImplementedError` + docs

Happy to do either; will figure out during implementation.

---

### Backcompat / migration

- **No public API removed.** Every existing call works unchanged.
- **No SemVer break.** All new kwargs are keyword-only with defaults; all new exports are additive.
- **`_hdf5_file=` is keyword-only and underscore-prefixed** — explicitly not public API. Third-party callers of `read_hdf5_dataset(path, dataset)` continue to work.
- **Bug fix:** `HDF5Video` is now picklable when `_open_reader` is populated. CHANGELOG entry.
- **Behavior change:** `Video.exists()` for URL filenames now does a HEAD probe (cached for 60 s). Previously returned `False` for URLs. CHANGELOG entry.

---

### Deferred to follow-up PRs

| PR | Scope | Notes |
|---|---|---|
| Remote video | `load_video(url)` via PyAV (`av.open(url, options=...)`) | Documented FFmpeg CVE-surface warning for remote video |
| Google Drive resolver | ~70 LOC port from `gdown` resolves Drive UI URLs → direct download URLs | Drive has its own quirks (25 dl/day cap, HEAD often 405, redirects to `googleusercontent.com`) |
| `save_slp(url)` | `fsspec.open(url, "wb")` for `s3://`/`gs://`/`az://` | Cheap because fsspec handles multipart upload. HTTP `PUT` for HDF5 remains permanently out — unworkable for the file format. |
| (Future) | Process-level shared fsspec instance cache, async API, obstore adapter | If/when measured need arises |

---

I'll open the PR (or PRs) on `feature/remote-url-loading-fsspec`.

**Comment by talmo** [comment]
Implemented and merged to `main` as a 3-PR stack: #439 (core URL + cloud loading via fsspec), #442 (remote video via PyAV), and #441 (Google Drive share links). `sio.load_slp("https://...")`, cloud schemes (`sleap-io[cloud]`), `sio.load_video(url)` (`sleap-io[pyav]`), and Google Drive links now work, including lazy embedded-frame streaming from remote `pkg.slp`. Cohesive user docs are tracked in #443.
